minio/minio
# Enable event persistence in kafka

## Description
Kafka events will be persisted in a queue dir if configured, else persisted in memory.

## Motivation and Context
To provide offline backup/persistence for events, so that no events are lost.

## Regression
Its a feature

## How Has This Been Tested?

- Start MinIO without an active kafka broker with the following config
```
{
        "enable": true,
        "brokers": ["localhost:9092", "localhost:9093", "localhost:9094"],
        "topic": "bucketevents",
        "queueDir": "/tmp/kafka",
        "queueLimit": 1000,
        "tls": {
            "enable": false,
            "skipVerify": false,
            "clientAuth": 0
        },
        "sasl": {
            "enable": false,
            "username": "",
            "password": ""
        }
}
```
- Enable bucket notification using mc
- Start sending events
- The events will persist in `/tmp/kafka` and gets replayed when the broker is turned on. (/tmp/kafka - will be empty if all the events are successfully sent - atleast leader is online)
- Set up a multi-node kafka cluster
- For an event to be successfully sent, the leader must be online. You can see the leader node by - `bin/kafka-topics.sh --describe --bootstrap-server localhost:9092 --topic bucketevents` 
- The events will be replayed from the store. which can be monitored by - `kafkacat -b localhost:9092 -t bucketevents`
- Bring the leader node up/down to eventually test the persistence and replay.

## Types of changes


## Checklist:
